Add support for int[n]? arguments in schema

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43387 centralize autograd dispatch key set
* **#43262 Add support for int[n]? arguments in schema**

Previously, we supported optional int lists (`int[]?`, added in #37174) and scalar-fill int lists (`int[n]`, which takes a passed bare int and creates a list of `n` copies). Here we add support for scalar-fill optional int lists (`int[n]?`) which combine the two behaviors.

Reviewer's note: the logic to handle the combination of optionality and scalar fill was already present in both eager and jit binding layers, this PR just 
* extends the existing legacy schema-tweaking logic in the codegen pipeline to make the necessary adjustments for python_arg_parser
* extends the JIT schema parser to handle extracting size information from the new form 

Once legacy schema support is no longer needed, the former should disappear. Also WRT to the former, @ezyang I wouldn't imagine this will be too hard to add to your rewrite but FYI.

Differential Revision: [D23212049](https://our.internmc.facebook.com/intern/diff/D23212049)

Commit:
add support for optional int list with scalar fill

[ghstack-poisoned]

Commit:
Update on "add support for optional int list with scalar fill"


Differential Revision: [D23212049](https://our.internmc.facebook.com/intern/diff/D23212049)

[ghstack-poisoned]

Commit:
Update on "Add support for int[n]? arguments in schema"


Previously, we supported optional int lists (`int[]?`, added in #37174) and scalar-fill int lists (`int[n]`, which takes a passed bare int and creates a list of `n` copies). Here we add support for scalar-fill optional int lists (`int[n]?`) which combine the two behaviors.

Reviewer's note: the logic to handle the combination of optionality and scalar fill was already present in both eager and jit binding layers, this PR just 
* teaches the existing legacy schema-tweaking logic in the codegen pipeline 
* extends the JIT schema parser
to handle the new combination. Once everybody's working from the JIT schema, the former should disappear. Also WRT to the former, @ezyang I wouldn't imagine this will be too hard to add to your rewrite but FYI.

Differential Revision: [D23212049](https://our.internmc.facebook.com/intern/diff/D23212049)

[ghstack-poisoned]

bhosmer commented:
@dreiss I was trying to remember if there were deep problems that dissuaded us from adding this in/around #37174 - IIRC it wasn't because it was impossible, just not compelling ROI given what you needed. But curious to know if I'm forgetting anything.

ezyang commented:
I agree this should be easy to port.

Commit:
Update on "Add support for int[n]? arguments in schema"


Previously, we supported optional int lists (`int[]?`, added in #37174) and scalar-fill int lists (`int[n]`, which takes a passed bare int and creates a list of `n` copies). Here we add support for scalar-fill optional int lists (`int[n]?`) which combine the two behaviors.

Reviewer's note: the logic to handle the combination of optionality and scalar fill was already present in both eager and jit binding layers, this PR just 
* extends the existing legacy schema-tweaking logic in the codegen pipeline to make the necessary adjustments for python_arg_parser
* extends the JIT schema parser to handle extracting size information from the new form 

Once legacy schema support is no longer needed, the former should disappear. Also WRT to the former, @ezyang I wouldn't imagine this will be too hard to add to your rewrite but FYI.

Differential Revision: [D23212049](https://our.internmc.facebook.com/intern/diff/D23212049)

[ghstack-poisoned]

Commit:
Update on "Add support for int[n]? arguments in schema"


Previously, we supported optional int lists (`int[]?`, added in #37174) and scalar-fill int lists (`int[n]`, which takes a passed bare int and creates a list of `n` copies). Here we add support for scalar-fill optional int lists (`int[n]?`) which combine the two behaviors.

Reviewer's note: the logic to handle the combination of optionality and scalar fill was already present in both eager and jit binding layers, this PR just 
* extends the existing legacy schema-tweaking logic in the codegen pipeline to make the necessary adjustments for python_arg_parser
* extends the JIT schema parser to handle extracting size information from the new form 

Once legacy schema support is no longer needed, the former should disappear. Also WRT to the former, @ezyang I wouldn't imagine this will be too hard to add to your rewrite but FYI.

Differential Revision: [D23212049](https://our.internmc.facebook.com/intern/diff/D23212049)

[ghstack-poisoned]

Commit:
Update on "Add support for int[n]? arguments in schema"


Previously, we supported optional int lists (`int[]?`, added in #37174) and scalar-fill int lists (`int[n]`, which takes a passed bare int and creates a list of `n` copies). Here we add support for scalar-fill optional int lists (`int[n]?`) which combine the two behaviors.

Reviewer's note: the logic to handle the combination of optionality and scalar fill was already present in both eager and jit binding layers, this PR just 
* extends the existing legacy schema-tweaking logic in the codegen pipeline to make the necessary adjustments for python_arg_parser
* extends the JIT schema parser to handle extracting size information from the new form 

Once legacy schema support is no longer needed, the former should disappear. Also WRT to the former, @ezyang I wouldn't imagine this will be too hard to add to your rewrite but FYI.

Differential Revision: [D23212049](https://our.internmc.facebook.com/intern/diff/D23212049)

[ghstack-poisoned]

dreiss commented:
Yeah.  The semantics of auto-conversion were slightly tricky, and I didn't want to bother because my callsite was already doing the expansion in Python (and couldn't easily adopt auto-expansion even if it were available).

Commit:
Update on "Add support for int[n]? arguments in schema"


Previously, we supported optional int lists (`int[]?`, added in #37174) and scalar-fill int lists (`int[n]`, which takes a passed bare int and creates a list of `n` copies). Here we add support for scalar-fill optional int lists (`int[n]?`) which combine the two behaviors.

Reviewer's note: the logic to handle the combination of optionality and scalar fill was already present in both eager and jit binding layers, this PR just 
* extends the existing legacy schema-tweaking logic in the codegen pipeline to make the necessary adjustments for python_arg_parser
* extends the JIT schema parser to handle extracting size information from the new form 

Once legacy schema support is no longer needed, the former should disappear. Also WRT to the former, @ezyang I wouldn't imagine this will be too hard to add to your rewrite but FYI.

Differential Revision: [D23212049](https://our.internmc.facebook.com/intern/diff/D23212049)

[ghstack-poisoned]
